File: src/test/java/ru/anseranser/peshki/engine/GameEngineTest.java
package ru.anseranser.peshki.engine;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import ru.anseranser.peshki.engine.event.GameEvent;
import ru.anseranser.peshki.input.MoveCommand;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class GameEngineTest {

    private GameConfig config;
    private GameEngine engine;

    @BeforeEach
    void setUp() {
        config = GameConfig.DEFAULT;
        engine = new GameEngine(config);
    }

    @Test
    void engineStartsWithZeroTurns() {
        assertEquals(0, engine.getTurnNumber());
    }

    @Test
    void engineHas4Players() {
        assertEquals(4, engine.getBoard().getPlayers().size());
    }

    @Test
    void firstPlayerIsHuman() {
        assertTrue(engine.getBoard().getPlayers().get(0).isHuman());
    }

    @Test
    void rollDiceReturnsTwoValues() {
        List<Integer> dice = engine.rollDice();
        assertEquals(2, dice.size());
        assertTrue(dice.stream().allMatch(d -> d >= 1 && d <= 6));
    }

    @Test
    void botTurnIncrementsTurnNumber() {
        engine.executeBotTurn();
        assertEquals(1, engine.getTurnNumber());
    }

    @Test
    void botTurnProducesTurnEndedEvent() {
        List<GameEvent> events = engine.executeBotTurn();
        assertTrue(events.stream().anyMatch(e -> e instanceof GameEvent.TurnEnded));
    }

    @Test
    void multipleBotTurnsAdvancePlayers() {
        engine.executeBotTurn();
        engine.executeBotTurn();
        engine.executeBotTurn();
        assertTrue(engine.getTurnNumber() >= 3);
    }

    @Test
    void generateAllMovesWithMovablePawns() {
        Player player = engine.getBoard().getPlayers().get(1);
        // Place a pawn on the field
        player.getPawns().getFirst().setCell(engine.getBoard().getCorner(player).getNextFieldCell());
        player.getPawns().getFirst().setState(Pawn.State.FIELDER);
        engine.getBoard().getCorner(player).getNextFieldCell().setPawn(player.getPawns().getFirst());

        List<Integer> dice = List.of(3, 4);
        List<Move> moves = GameEngine.generateAllMoves(player, dice, engine.getBoard(), config);
        assertFalse(moves.isEmpty());
    }

    @Test
    void generateAllMovesIncludesSumMoves() {
        Player player = engine.getBoard().getPlayers().get(1);
        player.getPawns().getFirst().setCell(engine.getBoard().getCorner(player).getNextFieldCell());
        player.getPawns().getFirst().setState(Pawn.State.FIELDER);
        engine.getBoard().getCorner(player).getNextFieldCell().setPawn(player.getPawns().getFirst());

        List<Integer> dice = List.of(3, 4);
        List<Move> moves = GameEngine.generateAllMoves(player, dice, engine.getBoard(), config);

        boolean hasSumMove = moves.stream()
                .anyMatch(m -> m.pawn() != null && m.steps() == 7);
        assertTrue(hasSumMove);
    }

    @Test
    void generateAllMovesWithNoMovablePawnsReturnsEmpty() {
        Player player = engine.getBoard().getPlayers().get(1);
        // All pawns on BENCH
        List<Integer> dice = List.of(3, 4);
        List<Move> moves = GameEngine.generateAllMoves(player, dice, engine.getBoard(), config);
        assertTrue(moves.isEmpty());
    }

    @Test
    void fullBotGameCompletes() {
        GameEngine botEngine = new GameEngine(config);
        botEngine.getBoard().getPlayers().forEach(p -> p.setHuman(false));

        int turns = 0;
        while (!botEngine.isGameOver() && turns < config.maxTurns()) {
            botEngine.executeBotTurn();
            turns++;
        }

        assertTrue(botEngine.isGameOver() || turns >= config.maxTurns());
    }

    @Test
    void placeCommandPlacesPawnWithoutRejection() {
        Player player = engine.getBoard().getPlayers().get(0);
        List<GameEvent> events = engine.executeHumanCommand(new MoveCommand.PlacePawn(1, 6));
        assertFalse(events.stream().anyMatch(e -> e instanceof GameEvent.MoveRejected),
                "Placing a pawn must not be rejected");
        assertTrue(events.stream().anyMatch(e -> e instanceof GameEvent.PawnPlaced));
    }

    @Test
    void placeAndMoveCommandWorksAndMovesPawn() {
        Player player = engine.getBoard().getPlayers().get(0);
        List<GameEvent> events = engine.executeHumanCommand(new MoveCommand.PlaceAndMove(1, 6, 2));

        assertFalse(events.stream().anyMatch(e -> e instanceof GameEvent.MoveRejected),
                "PlaceAndMove must not be rejected");
        assertTrue(events.stream().anyMatch(e -> e instanceof GameEvent.PawnPlaced));
        assertTrue(events.stream().anyMatch(e -> e instanceof GameEvent.PawnMoved),
                "The placed pawn must be moved after placement");

        boolean hasFielder = player.getPawns().stream()
                .anyMatch(p -> p.getState() == Pawn.State.FIELDER);
        assertTrue(hasFielder, "Placed pawn should have advanced onto the field");
    }

    @Test
    void plainMoveWithoutKillIsNotRejected() {
        Player player = engine.getBoard().getPlayers().get(0);
        engine.executeHumanCommand(new MoveCommand.PlacePawn(1, 6));

        Pawn placed = player.getPawns().stream()
                .filter(p -> p.getState() == Pawn.State.NEWBORN
                        && p.getCell() == engine.getBoard().getCorner(player))
                .findFirst()
                .orElse(null);
        assertNotNull(placed, "A pawn should be placed on the corner");

        List<GameEvent> move = engine.executeHumanCommand(
                new MoveCommand.MovePawn(1, placed.getNumber(), 1, List.of()));
        assertFalse(move.stream().anyMatch(e -> e instanceof GameEvent.MoveRejected),
                "A non-killing move must not be rejected");
        assertTrue(move.stream().anyMatch(e -> e instanceof GameEvent.PawnMoved));
    }
}

File: web/tests/GameEngine.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { DEFAULT_CONFIG } from '../src/engine/GameConfig';
import { GameEngine } from '../src/engine/GameEngine';
import { PawnState } from '../src/engine/Pawn';
import { GameEngine as GE } from '../src/engine/GameEngine';

describe('GameEngine', () => {
  let engine: GameEngine;

  beforeEach(() => {
    engine = new GameEngine(DEFAULT_CONFIG);
  });

  it('starts with zero turns', () => {
    expect(engine.turnNumber).toBe(0);
  });

  it('has 4 players', () => {
    expect(engine.board.players.length).toBe(4);
  });

  it('first player is human', () => {
    expect(engine.board.players[0].isHuman).toBe(true);
  });

  it('rollDice returns two values', () => {
    const dice = engine.rollDice();
    expect(dice.length).toBe(2);
    expect(dice.every(d => d >= 1 && d <= 6)).toBe(true);
  });

  it('bot turn increments turn number', () => {
    engine.executeBotTurn();
    expect(engine.turnNumber).toBe(1);
  });

  it('bot turn produces turn ended event', () => {
    const events = engine.executeBotTurn();
    expect(events.some(e => e.type === 'TurnEnded')).toBe(true);
  });

  it('multiple bot turns advance players', () => {
    engine.executeBotTurn();
    engine.executeBotTurn();
    engine.executeBotTurn();
    expect(engine.turnNumber).toBeGreaterThanOrEqual(3);
  });

  it('generateAllMoves with movable pawns', () => {
    const player = engine.board.players[1];
    player.pawns[0].cell = engine.board.getCorner(player).nextFieldCell;
    player.pawns[0].state = PawnState.FIELDER;
    engine.board.getCorner(player).nextFieldCell!.pawn = player.pawns[0];

    const dice = [3, 4];
    const moves = GE.generateAllMoves(player, dice, engine.board, DEFAULT_CONFIG);
    expect(moves.length).toBeGreaterThan(0);
  });

  it('generateAllMoves includes sum moves', () => {
    const player = engine.board.players[1];
    player.pawns[0].cell = engine.board.getCorner(player).nextFieldCell;
    player.pawns[0].state = PawnState.FIELDER;
    engine.board.getCorner(player).nextFieldCell!.pawn = player.pawns[0];

    const dice = [3, 4];
    const moves = GE.generateAllMoves(player, dice, engine.board, DEFAULT_CONFIG);

    const hasSumMove = moves.some(m => m.pawn !== null && m.steps === 7);
    expect(hasSumMove).toBe(true);
  });

  it('generateAllMoves with no movable pawns returns empty', () => {
    const player = engine.board.players[1];
    const dice = [3, 4];
    const moves = GE.generateAllMoves(player, dice, engine.board, DEFAULT_CONFIG);
    expect(moves.length).toBe(0);
  });

  it('placeAndMove command works', () => {
    const player = engine.board.players[0];
    const events = engine.executeHumanCommand(
      { type: 'PlacePawn', playerNumber: 1, diceValue: 6 },
      [],
    );
    expect(events.length).toBeGreaterThan(0);
    expect(events.some(e => e.type === 'PawnPlaced')).toBe(true);
  });

  it('full bot game completes', () => {
    const botEngine = new GameEngine(DEFAULT_CONFIG);
    botEngine.board.players.forEach(p => { p.isHuman = false; });

    let turns = 0;
    while (!botEngine.isGameOver() && turns < DEFAULT_CONFIG.maxTurns) {
      botEngine.executeBotTurn();
      turns++;
    }

    expect(botEngine.isGameOver() || turns >= DEFAULT_CONFIG.maxTurns).toBe(true);
  });
});
